### PR TITLE
clientv3: version checking

### LINKED ISSUE #6579: client versioning
If a new client is used with an old cluster, it may try to use options or flags the old server does not understand. Instead, there could be a `Config` option (or make it mandatory) to compare the client version against the cluster version; if the cluster is behind the client, the client will refuse to start.

Related: #6422

### PR DESCRIPTION
**Pull request by heyitsanthony**
Adds a `UseLatestCluster` flag (and user-defined context) to config.

Fixes #6579

### CODE REVIEW

**Comment by gyuho** [comment]
So this only checks against clients <3.2 trying to connect to 3.2> with this flag on, right?

File: clientv3/client.go [hunk 1]
```diff
@@ -353,10 +360,57 @@ func newClient(cfg *Config) (*Client, error) {
...
+	}
+
 	go client.autoSync()
 	return client, nil
 }
 
+func (c *Client) checkVersion() (err error) {
+	var wg sync.WaitGroup
+	errc := make(chan error, len(c.cfg.Endpoints))
+	ctx, cancel := context.WithCancel(c.ctx)
```
Comment (gyuho) [inline clientv3/client.go]: Does `cancel` func preserver even when we overwrite `ctx`?

Comment (heyitsanthony) [inline clientv3/client.go]: yeah, canceling the parent context will cancel any descendents

Comment (gyuho) [inline clientv3/client.go]: Oh I see it now `context.WithTimeout(ctx`, not `context.WithTimeout(c.ctx`.

Makes sense.

File: clientv3/config.go [hunk 2]
```diff
@@ -41,6 +42,13 @@ type Config struct {
 	// Password is a password for authentication.
 	Password string `json:"password"`
 
+	// UseLatestCluster when set will refuse to create a client against an outdated cluster.
+	UseLatestCluster bool `json:"use-latest-cluster"`
```
Comment (xiang90) [inline clientv3/config.go]: EnsureClusterVersion? Or CheckClusterVersion? `use` is not a strong enough word I feel.

Comment (heyitsanthony) [inline clientv3/config.go]: `RejectOldCluster`? Should have something about whether it's new/old/latest

File: clientv3/client.go [hunk 3]
```diff
@@ -353,10 +360,57 @@ func newClient(cfg *Config) (*Client, error) {
...
+				return
+			}
+			vs := strings.Split(resp.Version, ".")
+			maj, min := 0, 0
+			if len(vs) >= 2 {
+				maj, rerr = strconv.Atoi(vs[0])
+				min, rerr = strconv.Atoi(vs[1])
+			}
+			if maj < 3 || min < 2 {
+				rerr = rpctypes.ErrNotCapable
```
Comment (xiang90) [inline clientv3/client.go]: shall we define a different error for cluster version mismatch?

File: clientv3/config.go (hunk 2 above)
Comment (xiang90) [inline clientv3/config.go]: RejectOldCluster looks good.

File: clientv3/client.go (hunk 3 above)
Comment (heyitsanthony) [inline clientv3/client.go]: On the fence about it. I feel it's similar enough to `NotCapable` (i.e., a "soft" not capable) that the error can be reused.

Comment (xiang90) [inline clientv3/client.go]: NotCapable is the error returned from server (at least for its current definition). The string is `"etcdserver: not capable"`. It has the server prefix. But here users would expect `client: server is not capable`.

Comment (heyitsanthony) [inline clientv3/client.go]: OK, fair enough

**Comment by heyitsanthony** [comment]
all fixed ptal

**Comment by gyuho** [comment]
LGTM. /cc @xiang90

**Comment by xiang90** [comment]
LGTM

**Pull request merged**